### ISSUE TITLE
Application state

### ISSUE THREAD

**Issue by daryl** [lhorie/mithril.js#423]
I'm using Browserify (CommonJS) to build my app and I'm wondering about the best way to handle an application state.

For example, let's say I have multiple routes:

```js
import m from "mithril"

var dummy = function() {
  var d = m.deferred()
  m.startComputation()
  setTimeout(function() {
    d.resolve()
    m.endComputation()
  }, Math.random()*3000)
  return d.promise
}

var App = {
  controller: function() {
    this.obj = dummy()
  },
  view: function(ctrl) {
    m("a", {path: "/foo", config: m.route}, "Foo")
  }
}

var Foo = {
  controller: function() {
    this.obj = dummy()
  },
  view: function(ctrl) {
    m("a", {path: "/", config: m.route}, "Root")
  }
}

m.route.mode = "pathname"

m.route(document.body, "/", {
  "/": App,
  "/foo": Foo
})
```

What's the best way to add a loading class to the `body` element whilst routing and for it to remove once it's been redrawn? If I navigate to `Foo` from `App`, and `Foo` makes an AJAX call which takes, hypothetically, 2 seconds to return, that's a long time and the user may question whether the application is responding at all.

**Comment by lhorie** [comment]
There's a couple of ways to add a loading icon. The cheapest is shown in [this article](http://lhorie.github.io/mithril-blog/wait-for-it.html).

Basically, the gist is to simply add the class via plain DOM manipulation before the request and remove it afterwards.

There are some other suggestions in the comments for the article as well.

You can configure requests so that redrawing occurs both before AND after the request (as opposed to only after):

```javascript
ctrl.loading = true
m.request({method: "GET", url: "/api/foo", background: true, initialValue: []})
  .then(function(value) {
    //do stuff
    ctrl.loading = false
  })
```

**Comment by barneycarroll** [comment]
Hey @daryl, I've got loads of stuff loading on my page all the time. The way I handle it is to create a model object for the resource I'm loading before it is requested, and extend it with the resource data afterwards. As @lhorie describes, I set a `loading` property to `true` and call `m.redraw()` just before loading, then set it to `false` in the success callback (I also set an `error` property in the failure callback, but hopefully you don't have that kind of problem ;).

If you can post some pseudo-code with your existing model I'd be happy to advise on more detail.

**Comment by daryl** [comment]
That's perfect, thanks for the help, guys.
